contracts/CompoundFCM.sol

```diff
@@ -0,0 +1,337 @@
+
+    /// @audit-casting variableTokenDelta is expected to be positive here, but what if goes below 0 due to rounding imprecision?
+    uint256 updatedTraderMargin = trader.marginInScaledYieldBearingTokens - uint256(variableTokenDelta).wadDiv(currentExchangeRate);
+    trader.updateMarginInScaledYieldBearingTokens(updatedTraderMargin);
+
+    // check the margin requirement of the trader post unwind, if the current balances still support the unwind, they it can happen, otherwise the unwind will get reverted
+    checkMarginRequirement(_fixedTokenBalance, _variableTokenBalance, trader.marginInScaledYieldBearingTokens);
+
+    // transfer fees to the margin engine
+    underlyingToken.safeTransferFrom(msg.sender, address(_marginEngine), cumulativeFeeIncurred);
+
+    // transfer the yield bearing tokens to trader address and update margin in terms of yield bearing tokens
+    // variable token delta should be positive
+    IERC20Minimal(address(_ctoken)).safeTransfer(msg.sender, uint256(variableTokenDelta));
+
+  }
+
+
+  /// @notice Check Margin Requirement post unwind of a fully collateralised fixed taker
+  function checkMarginRequirement(int256 traderFixedTokenBalance, int256 traderVariableTokenBalance, uint256 traderMarginInScaledYieldBearingTokens) internal {
+
+    // variable token balance should never be positive
+    // margin in scaled tokens should cover the variable leg from now to maturity
+
+    /// @dev we can be confident the variable token balance of a fully collateralised fixed taker is always going to be negative (or zero)
+    /// @dev hence, we can assume that the variable cashflows from now to maturity is covered by a portion of the trader's collateral in yield bearing tokens
+    /// @dev once future variable cashflows are covered, we need to check if the remaining settlement cashflow is covered by the remaining margin in yield bearing tokens
+
+    // @audit: casting variableTokenDelta is expected to be positive here, but what if goes below 0 due to rounding imprecision?
+    uint256 marginToCoverVariableLegFromNowToMaturity = uint256(-traderVariableTokenBalance);
+    int256 marginToCoverRemainingSettlementCashflow = int256(getTraderMarginInYieldBearingTokens(traderMarginInScaledYieldBearingTokens)) - int256(marginToCoverVariableLegFromNowToMaturity);
+
+    int256 remainingSettlementCashflow = calculateRemainingSettlementCashflow(traderFixedTokenBalance, traderVariableTokenBalance);
+
+    if (remainingSettlementCashflow < 0) {
+
+      if (-remainingSettlementCashflow > marginToCoverRemainingSettlementCashflow) {
+        revert CustomErrors.MarginRequirementNotMetFCM(int256(marginToCoverVariableLegFromNowToMaturity) + remainingSettlementCashflow);
+      }
+
+    }
+
+  }
+
+
+  /// @notice Calculate remaining settlement cashflow
+  function calculateRemainingSettlementCashflow(int256 traderFixedTokenBalance, int256 traderVariableTokenBalance) internal returns (int256 remainingSettlementCashflow) {
```

Comment:
is this function a duplicate of the function in AaveFCM ? if yes, worth pushing into a separate library?

Reply:
I think we should probably move it into a BaseFCM contract, along with other common concepts like vamm, underlying, rateOracle and marginEngine (which this function will anyway need access to). But that will be quite destabilising to the existing AaveFCM and is quite a big change to be making on the same day we are sending for audit. Do we want to do this?

Reply:
not worth doing this for the audit, but worth putting this into the rate oracle optimisation ticket as one of many items that could help us optimise the zero to one process for new rate oracles

Reply:
👍  Agreed